docs/designs/termination.md

```diff
@@ -0,0 +1,130 @@
+
+After all pods evict, we will attempt to terminate the node. If the cloud-provider API returns an error, preventing instance termination, we will log the error in the Karpenter pod and node status then exponentially back off and retry.
+
+In the case where a user has enabled termination protection for their underlying instance, Karpenter will be unable to terminate the instance until the termination protection is removed. Since we only terminate instances corresponding to  Karpenter provisioned nodes, a user would have to manually set this protection after Karpenter creates it. In this case, we depend on the user to resolve this with the node status and Karpenter logs.
+
+### User Configuration
+
+The termination controller and associated webhooks will come installed with Karpenter, requiring no additional configuration on the user’s part. 
+
+We will allow users to specify a `karpenter.sh/do-not-evict` label on their pods, guaranteeing that we will not evict certain pods. A node with `do-not-evict` pods will cordon but wait to drain until all `do-not-evict` pods are gone. This way, the cluster will utilize its existing allocated capacity until the `do-not-evict` pods terminate. Users can use Karpenter’s scheduling logic to colocate pods with this label onto similar nodes to load balance these pods.
+
+### Parallel Node Termination
+
+The termination controller will be able to drain and terminate multiple nodes in parallel. We will expedite this process while rate limiting node termination to protect against throttling.
+
+We introduce an optional cluster-scoped CRD, the Node Disruption Budget (NDB), a [Pod Disruption Budget (PDB)](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-budgets) for nodes. A user can scope an NDB to Karpenter provisioned nodes through a label selector, since nodes are created with their Provisioner’s labels. `Unavailable` nodes will be `NotReady` or have `metadata.DeletionTimestamp` set. `Available` nodes will be `Ready`.
+
+A termination is allowed if at least minAvailable nodes selected by selector will still be available after the termination. For example, you can prevent all terminations by specifying “100%”. A termination is also allowed if at most maxUnavailable nodes selected by selector are unavailable after the termination. For example, one can prevent all terminations by specifying 0. These two are mutually exclusive.  
```

Comment:
nit: "by a selector"

"maxUnavailable nodes selected by a selector"

Can you also say the parameter names "minAvailable" and "maxUnavailable" are the mutually exclusive parameters rather than just saying "These two"?

Reply:
Thanks again. Good idea here, better to be explicit.